### PR TITLE
feat(zk): implement halo2 prover

### PR DESCRIPTION
**Pull request by chokobole**
# Description

`Halo2Prover` is a child clas of `Prover` and it holds commonly used `PolynomicalCommitmentScheme`, `Domain`, `Blinder`, `TranscriptWriter` and `Halo2RandomFieldGenerator`.

Should be merged after #153

### CODE REVIEW

File: tachyon/zk/base/halo2_prover.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      std::unique_ptr<TranscriptWriter<Commitment>> writer,
+      std::unique_ptr<crypto::XORShiftRNG> rng, size_t blinding_factors) {
+    auto generator = std::make_unique<Halo2RandomFieldGenerator<F>>(rng.get());
+    Blinder<PCSTy> blinder(generator.get(), blinding_factors);
+    return {std::move(pcs),    std::move(domain), std::move(blinder),
+            std::move(writer), std::move(rng),    std::move(generator)};
+  }
+
+  crypto::XORShiftRNG* rng() { return rng_.get(); }
+  Halo2RandomFieldGenerator<F>* generator() { return generator_.get(); }
```
Comment (Insun35) [inline tachyon/zk/base/halo2_prover.h:57]:
```suggestion
  crypto::XORShiftRNG* rng() const { return rng_.get(); }
  Halo2RandomFieldGenerator<F>* generator() const { return generator_.get(); }
```

Comment (chokobole) [reply tachyon/zk/base/halo2_prover.h:57]: `const` should not be attached here. Otherwise you can't do anything with `rng()` or `generator()`

**Review by TomTaehoonKim — APPROVED** [review]
Can you just fix typo in the commit message? clas->class.
Else, looks good to me!

File: tachyon/math/finite_fields/generator/prime_field_generator/prime_field_generator.cc
Comment (chokobole) [inline tachyon/math/finite_fields/generator/prime_field_generator/prime_field_generator.cc:1]: I just added a comment to not merge before #153

**Review by Insun35 — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Pull request merged**